core/src/main/java/org/apache/brooklyn/core/config/internal/AbstractConfigMapImpl.java

```diff
@@ -20,65 +20,212 @@
 
-public abstract class AbstractConfigMapImpl implements ConfigMap {
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+
+public abstract class AbstractConfigMapImpl<TContainer extends BrooklynObject> implements ConfigMapWithInheritance<TContainer> {
+
+    /*
+     * Changed Sept 2016 so that keys can determine inheritance strategy at every level in the hierarchy,
+     * and signifcantly refactored to share code among subclasses, adding Location as a subclass.
+     */
 
     private static final Logger LOG = LoggerFactory.getLogger(AbstractConfigMapImpl.class);
     
-    protected final ConfigMapViewWithStringKeys mapViewWithStringKeys = new ConfigMapViewWithStringKeys(this);
+    @SuppressWarnings("deprecation")
+    protected final transient org.apache.brooklyn.core.entity.internal.ConfigMapViewWithStringKeys mapViewWithStringKeys = new org.apache.brooklyn.core.entity.internal.ConfigMapViewWithStringKeys(this);
+
+    protected TContainer bo;
     
     /**
      * Map of configuration information that is defined at start-up time for the entity. These
      * configuration parameters are shared and made accessible to the "children" of this
      * entity.
      */
-    protected Map<ConfigKey<?>,Object> ownConfig = Collections.synchronizedMap(new LinkedHashMap<ConfigKey<?>, Object>());
+    protected final Map<ConfigKey<?>,Object> ownConfig;
+    
+    protected AbstractConfigMapImpl(TContainer bo) {
+        // Not using ConcurrentMap, because want to (continue to) allow null values.
+        // Could use ConcurrentMapAcceptingNullVals (with the associated performance hit on entrySet() etc).
+        this(bo, Collections.synchronizedMap(new LinkedHashMap<ConfigKey<?>, Object>()));
+    }
+    protected AbstractConfigMapImpl(TContainer bo, Map<ConfigKey<?>, Object> storage) {
+        this.bo = bo;
+        this.ownConfig = storage;
+    }
 
+    public TContainer getContainer() {
+        return bo;
+    }
+    
+    protected final BrooklynObjectInternal getBrooklynObject() {
+        return (BrooklynObjectInternal)bo;
+    }
+    
     public <T> T getConfig(ConfigKey<T> key) {
-        return getConfig(key, null);
+        return getConfigImpl(key, false).getWithoutError().get();
     }
     
     public <T> T getConfig(HasConfigKey<T> key) {
-        return getConfig(key.getConfigKey(), null);
+        return getConfigImpl(key.getConfigKey(), false).getWithoutError().get();
     }
     
-    public <T> T getConfig(HasConfigKey<T> key, T defaultValue) {
-        return getConfig(key.getConfigKey(), defaultValue);
+    @Override
+    public Maybe<Object> getConfigLocalRaw(ConfigKey<?> key) {
+        return getConfigRaw(key, false);
+    }
+
+    protected abstract ExecutionContext getExecutionContext(BrooklynObject bo);
+    protected abstract void postLocalEvaluate(ConfigKey<?> key, BrooklynObject bo, Maybe<?> rawValue, Maybe<?> resolvedValue);
```

Comment:
@ahgittin It wasn't obvious to me because nothing calls this method. I see from `EntityConfigMap.postLocalEvaluate` what the purpose is.

Is there a missing call to this method?